apps/dashboard/app/apps/nav_bar.rb

```diff
@@ -111,20 +112,22 @@ def self.item_from_token(token)
     end
   end
 
-  def self.extend_group(item)
-    NavItemDecorator.new(item, 'layouts/nav/group')
+  def self.extend_group(group)
+    group.sort = false
+    NavItemDecorator.new(group, 'layouts/nav/group', group.apps)
   end
 
-  def self.extend_link(item)
-    NavItemDecorator.new(item, 'layouts/nav/link')
+  def self.extend_link(link, apps: nil)
+    NavItemDecorator.new(link, 'layouts/nav/link', apps)
   end
 
   class NavItemDecorator < SimpleDelegator
-    attr_reader :partial_path
+    attr_reader :partial_path, :apps
 
-    def initialize(nav_item, partial_path)
+    def initialize(nav_item, partial_path, apps)
       super(nav_item)
       @partial_path = partial_path
+      @apps = apps
```

Comment:
Did some refactoring based on your comments and improved `BatchConnectConcern.bc_custom_apps_group`

Reply:
Thanks, I'm happy to defer it just the same. But something is (was) off right?

Reply:
Yes, it is a bit off.
I am much happier with the new implementation for the `BatchConnectConcern.bc_custom_apps_group`.
Still, adding the `apps` field to the nav_bar items has room for improvements. Something for a later date?

Reply:
>  Something for a later date?

Yes we can do it later but it's likely just in the next few weeks 😆.

Reply:
The other idea I had initially was to create a new class `InteractiveAppsMenu` that would parse the `nav_bar` config and create a navigation group based on the apps configured, then use it in the same way in `BatchConnectConcern.bc_custom_apps_group`. Very similar to `NavBar` class.

What do you think?

Reply:
I'm reading between the lines - `duplication is better than the wrong abstraction`.  If you want to move that way it may be OK, but still not quite right?

It seems like we want some sort of pipe and filter pattern where we start with everything and filter it to the subset you want to show.

Something like
```
navbar = all_apps.select(&:show_in_nav?)
menu = all_apps.select(&:show_in_menu?)
```

But there are a few complications given how we use & treat all these things. apps, bc apps & plain links don't have a unified interface.  Indeed apps and batch_connect apps have some relationship where they now about each other (instead of one way knowledge).

It seems like we need some sort of structure where they all look and feel like the _same thing_ even if they aren't.  But in any case, it's really a task (likely several) onto itself to get a handle on paying technical debt from years ago (apps & bc apps) and technical debt we're now incurring.

Reply:
I find this filter pattern a bit more difficult to configure.

If I got your idea correctly and to support having a custom navigation per profile, we need to read the default apps from the file system and then merge their navigation metadata with the one in the config.
eg: `sys/bc_jupyter` `manifest` has `show_in_nav=true` and `category='My Apps'`
but then the profile configuration can set `show_in_nav=false` or  `category='Custom category'` for that individual app to control how it appears in the navigation.

As for the unified interface, I think it would be best to create a new class that represents a NavItem and create mappers for each type that can be an item in the navigation.

I think a quick Slack Huddle tomorrow might be a good idea to discuss the best approach to move forward.

Reply:
We don't need to solve it now. I'm happy to move forward with this and refactor later.

Reply:
I have an idea after reviewing your comments and the current implementation.

It removes the need to add the `apps` field, but requires to expand all links when parsing the configuration to create the `NavBar`.

Let me know what do you think:
https://github.com/OSC/ondemand/pull/2374/commits/69a9dd0c8435d7b1f5ee5d0cfddd460882f139d6